---
title: "Selection_Analysis_Parsing"
author: "Zachary Kileeg"
date: "2/18/2021"
output: html_document
---
##chunk takes a directory of json files from FUBAR selection output, a directory with hmmer files to split into kinase/not kinase, 
###and an output. This will output the selected sites for each OG and a summary file 

###The second chunk contains both this cod eand the normalization code. Doesn't make sense to run both tbh 
```{r}
library(rjson)
library(dplyr)
library(stringr)
library(rhmmer)

selection_dir = "D:/Analysis_Output/hyphy/CDS/Dec_stuff/RLKs_only/Fubar_out/"
hmmer_dir = "D:/Analysis_Output/hyphy/CDS/Dec_stuff/RLKs_only/Translated/AlteredGapChar/Consensus/Pkinases/"
output_dir = "D:/Analysis_Output/hyphy/CDS/Dec_stuff/RLKs_only//Selection_Parsed/"

input_files = list.files(selection_dir, pattern="*.json$", full.names=FALSE)
input_pkinase_files = list.files(hmmer_dir, pattern="*Pkinase.hmm.out.txt$", full.names=FALSE)
input_pkinase_tyr_files = list.files(hmmer_dir, pattern="*Pkinase_Tyr.hmm.out.txt$", full.names=FALSE)

output_matrix = matrix(nrow=length(input_files), ncol=10)
colnames(output_matrix) = c("Gene", "Num_Pos_Sites", "Num_Neg_Sites", "Ratio", "Pos_ECD", "Neg_ECD", "ECD_Ratio", "Pos_Kinase", "Neg_Kinase", "Kinase_Ratio")

filter_cutoff = 0.90

#make matrix to hold the number of sites under selection. We're going to use this to normalize after. 
#num_sites_under_selection = matrix(nrow=length(input_files), ncol=1)
#rownames(num_sites_under_selection) = input_files



#thingy = read_domtblout(paste(hmmer_dir,input_pkinase_files[i], sep=""))

#test = read.table(paste(hmmer_dir, input_pkinase_files[i], sep=""))


                     
for (i in 1:length(input_files)){
  
  
  
  hmmer_pkinase = try(read.table(paste(hmmer_dir,input_pkinase_files[i], sep="")),silent=TRUE)
  hmmer_kinase_tyr = try(read.table(paste(hmmer_dir,input_pkinase_tyr_files[i], sep="")), silent=TRUE)
  
  
  if (class(hmmer_pkinase) == "try-error" && class(hmmer_kinase_tyr) == "try-error") {
    
      warning("Input HMM files are empty")
    
    kinase_positions = "Error"
    
  } else if (class(hmmer_pkinase) == "try-error" && class(hmmer_kinase_tyr) != "try-error"){
    
     colnames(hmmer_kinase_tyr) = c("target_name", "accession_species", "tlen", "domain_name", "accession_hmm", "qlen", "E-value",  "sequence_score",  "sequence_bias", "sequence_num", "seq_of", "c-Evalue",  "i-Evalue",  "domain_score",  "domain_bias",  "hmm_from",    "hmm_to",  "ali_from",    "ali_to",  "env_from",     "env_to",  "acc", "description_of_target")
     
    kinase_positions = as.data.frame(hmmer_kinase_tyr)
    kinase_positions$domain_name=str_remove(kinase_positions$domain_name,"\\..*")
    
  } else if (class(hmmer_pkinase) != "try-error" && class(hmmer_kinase_tyr) == "try-error") {
    
    colnames(hmmer_pkinase) = c("target_name", "accession_species", "tlen", "domain_name", "accession_hmm", "qlen", "E-value",  "sequence_score",  "sequence_bias", "sequence_num", "seq_of", "c-Evalue",  "i-Evalue",  "domain_score",  "domain_bias",  "hmm_from",    "hmm_to",  "ali_from",  "ali_to",  "env_from", "env_to",  "acc", "description_of_target")
    
    kinase_positions = as.data.frame(hmmer_pkinase)
    kinase_positions$domain_name=str_remove(kinase_positions$domain_name,"\\..*")
    
  } else if (class(hmmer_pkinase) != "try-error" && class(hmmer_kinase_tyr) != "try-error") {
    
  colnames(hmmer_pkinase) = c("target_name", "accession_species", "tlen", "domain_name", "accession_hmm", "qlen", "E-value",  "sequence_score",  "sequence_bias", "sequence_num", "seq_of", "c-Evalue",  "i-Evalue",  "domain_score",  "domain_bias",  "hmm_from",    "hmm_to",  "ali_from",  "ali_to",  "env_from", "env_to",  "acc", "description_of_target")
  
   
  colnames(hmmer_kinase_tyr) = c("target_name", "accession_species", "tlen", "domain_name", "accession_hmm", "qlen", "E-value",  "sequence_score",  "sequence_bias", "sequence_num", "seq_of", "c-Evalue",  "i-Evalue",  "domain_score",  "domain_bias",  "hmm_from",    "hmm_to",  "ali_from",    "ali_to",  "env_from",     "env_to",  "acc", "description_of_target")
  
  hmmer_pkinase = as.data.frame(hmmer_pkinase)
  hmmer_kinase_tyr = as.data.frame(hmmer_kinase_tyr)
  
  #hmmer_pkinase = read_domtblout(paste(hmmer_dir,input_pkinase_files[i], sep=""))

  #hmmer_kinase_tyr = read_domtblout(paste(hmmer_dir,input_pkinase_tyr_files[i],sep=""))
  
  extra_tyr = hmmer_kinase_tyr %>% filter(!(target_name %in% hmmer_pkinase$target_name))
  
  kinase_positions = rbind(hmmer_pkinase, extra_tyr)
  kinase_positions$domain_name=str_remove(kinase_positions$domain_name,"\\..*")
  
  }
  
  
  if (kinase_positions != "Error"){
  #gene_name = str_replace(basename(input_files[i]), "_.*","")
  gene_name= input_files[i]
  
  #kinase_start_pos = as.vector(kinase_positions[grep(gene_name,kinase_positions$domain_name),'env_from'])
  kinase_start_pos = as.vector(min(kinase_positions$env_from))
  
  #get json of selection 
  selection_json = rjson::fromJSON(file=paste(selection_dir,input_files[i],sep=""))
  
  if (selection_json$MLE$content$`0`[1]!="NULL"){
    
    selection_df = as.data.frame(t(as.data.frame(selection_json$MLE$content$`0`)))
    rownames(selection_df)=c(1:nrow(selection_df))
    selection_df = cbind(selection_df, rownames(selection_df)) 
    colnames(selection_df) = c("alpha", "beta", "beta-alpha, mean posterior beta-alpha", "negative", "positive", "bayes_factor", "null1", "null2", "Position")
   
    #for this selection file, find out how many sites were under selection for normalization
    #num_sites_under_selection[i,1] = selection_json$input$`number of sites`

    pos_sites = selection_df %>% filter(positive > filter_cutoff)
    neg_sites = selection_df %>% filter(negative > filter_cutoff)
    
    selection_ratio = nrow(pos_sites)/nrow(neg_sites)
    
    
    ###kinase filtering section#####
    
    ecd_sites = selection_df[1:kinase_start_pos[[1]],]
    kinase_sites = selection_df[kinase_start_pos[[1]]:nrow(selection_df),]  
    
    pos_ecd = ecd_sites %>% filter(positive > filter_cutoff)
    neg_ecd = ecd_sites %>% filter(negative > filter_cutoff)  
    
    pos_kinase = kinase_sites %>% filter(positive > filter_cutoff)
    neg_kinase = kinase_sites %>% filter(negative > filter_cutoff) 
    
    ecd_pos_neg_ratio = nrow(pos_ecd)/nrow(neg_ecd)
    kinase_pos_neg_ratio = nrow(pos_kinase)/nrow(neg_kinase)
    
    
     #get the average nonsyn rate and syn rate across the genes being compared
    
    syn_sites = mean(selection_df$alpha)
    
    nonsyn_sites = mean(selection_df$beta)
    
    #split the syn rates up by feature
    #ecd
    syn_ecd = mean(ecd_sites$alpha)
    nonsyn_ecd = mean(ecd_sites$beta)
    
    #kinase
    syn_kinase = mean(kinase_sites$alpha)
    nonsyn_kinase = mean(kinase_sites$beta)
    
    substitution_rate_matrix[i, 1:6] = c(syn_sites/gene_length, nonsyn_sites/gene_length, syn_ecd/ecd_length, nonsyn_ecd/ecd_length, syn_kinase/kinase_length, nonsyn_kinase/kinase_length)
    
    
    
    output_matrix[i,1:10] = c(gene_name, nrow(pos_sites), nrow(neg_sites), selection_ratio, nrow(pos_ecd), nrow(neg_ecd), ecd_pos_neg_ratio, nrow(pos_kinase), nrow(neg_kinase), kinase_pos_neg_ratio)
    
    
    
    
    
  } else {
    
    pos_sites = matrix(nrow=0, ncol=9)
    colnames(pos_sites) = c("alpha", "beta", "beta-alpha, mean posterior beta-alpha", "negative", "positive", "bayes_factor", "null1", "null2", "Position")
    
    neg_sites = matrix(nrow=0, ncol=9)
    colnames(neg_sites) = c("alpha", "beta", "beta-alpha, mean posterior beta-alpha", "negative", "positive", "bayes_factor", "null1", "null2", "Position")
    
    
    
    output_matrix[i,1:10] = c(gene_name, 0, 0, 0, 0, 0, 0, 0, 0, 0)
     
     
    }
  
  }
  
    
  
   
  
  
  #output positively selected sites
    write.csv(pos_sites, paste(output_dir, input_files[i], "_posSelection.csv", sep=""))
  
  #output negatively selected sites
   write.csv(neg_sites, paste(output_dir, input_files[i], "_negSelection.csv", sep=""))
  
}

#output total
write.csv(output_matrix, paste(output_dir, "RLKs_ALL.csv", sep=""), row.names=FALSE)
#write.csv(output_matrix, paste(output_dir, "RLKs_ALL_NormalizedL.csv", sep=""), row.names=FALSE)






```



#normalized     ####run this one to output both normalized and regular. The one previously is for just straight up parsed sites. No reason to run the previous one tbh, this one does both at the same time 
```{r}
library(rjson)
library(dplyr)
library(stringr)
library(rhmmer)
library(phylotools)


selection_dir = "D:/Analysis_Output/hyphy/CDS/Dec_stuff/RLKs_only/Fubar_out/"
hmmer_dir = "D:/Analysis_Output/hyphy/CDS/Dec_stuff/RLKs_only/Translated/AlteredGapChar/Consensus/Pkinases/"
output_dir = "D:/Analysis_Output/hyphy/CDS/Dec_stuff/RLKs_only//Selection_Parsed/"

input_files = list.files(selection_dir, pattern="*.json$", full.names=FALSE)
input_pkinase_files = list.files(hmmer_dir, pattern="*Pkinase.hmm.out.txt$", full.names=FALSE)
input_pkinase_tyr_files = list.files(hmmer_dir, pattern="*Pkinase.Tyr.hmm.out.txt$", full.names=FALSE)


#initialize the output matrices 
output_matrix = matrix(nrow=length(input_files), ncol=10)
colnames(output_matrix) = c("Gene", "Num_Pos_Sites", "Num_Neg_Sites", "Ratio", "Pos_ECD", "Neg_ECD", "ECD_Ratio", "Pos_Kinase", "Neg_Kinase", "Kinase_Ratio")

output_matrix_normalized = output_matrix

#initialize output matrix for substition rate matrices gene wide
substitution_rate_matrix = matrix(nrow=length(input_files), ncol=6)
colnames(substitution_rate_matrix) = c("Syn_rate", "NonSyn_Rate", "Syn_ECD", "NonSyn_ECD", "Syn_Kinase", "NonSyn_Kinase")
rownames(substitution_rate_matrix) = input_files

substitution_rate_matrix_normalized = substitution_rate_matrix

filter_cutoff = 0.90    ####this is a decimal posterior probability. max is 1. 



for (i in 1:length(input_files)){
  
   print(paste("Selection:", input_files[i], " hmmerfile", input_pkinase_files[i], sep=""))
  
  #try to read the kinase table into memory. If it doesn't work, it means most likely the file is empty. Catch this as an error
  hmmer_pkinase = try(read.table(paste(hmmer_dir,input_pkinase_files[i], sep="")),silent=TRUE)
  hmmer_kinase_tyr = try(read.table(paste(hmmer_dir,input_pkinase_tyr_files[i], sep="")), silent=TRUE)
  # 
  # 
  # if (class(hmmer_pkinase) != "try-error" && class(hmmer_kinase_tyr) != "try-error") {
  # colnames(hmmer_pkinase) = c("target_name", "accession_species", "tlen", "domain_name", "accession_hmm", "qlen", "E-value",  "sequence_score",  "sequence_bias", "sequence_num", "seq_of", "c-Evalue",  "i-Evalue",  "domain_score",  "domain_bias",  "hmm_from",    "hmm_to",  "ali_from",  "ali_to",  "env_from", "env_to",  "acc", "description_of_target")
  # 
  #  
  # colnames(hmmer_kinase_tyr) = c("target_name", "accession_species", "tlen", "domain_name", "accession_hmm", "qlen", "E-value",  "sequence_score",  "sequence_bias", "sequence_num", "seq_of", "c-Evalue",  "i-Evalue",  "domain_score",  "domain_bias",  "hmm_from",    "hmm_to",  "ali_from",    "ali_to",  "env_from",     "env_to",  "acc", "description_of_target")
  # 
  # hmmer_pkinase = as.data.frame(hmmer_pkinase)
  # hmmer_kinase_tyr = as.data.frame(hmmer_kinase_tyr)
  # 
  # #hmmer_pkinase = read_domtblout(paste(hmmer_dir,input_pkinase_files[i], sep=""))
  # 
  # #hmmer_kinase_tyr = read_domtblout(paste(hmmer_dir,input_pkinase_tyr_files[i],sep=""))
  # 
  # extra_tyr = hmmer_kinase_tyr %>% filter(!(target_name %in% hmmer_pkinase$target_name))
  # 
  # kinase_positions = rbind(hmmer_pkinase, extra_tyr)
  # kinase_positions$domain_name=str_remove(kinase_positions$domain_name,"\\..*")
  # 
  # hmmer_pkinase = try(read.table(paste(hmmer_dir,input_pkinase_files[i], sep="")),silent=TRUE)
  # hmmer_kinase_tyr = try(read.table(paste(hmmer_dir,input_pkinase_tyr_files[i], sep="")), silent=TRUE)
  
  
  if (class(hmmer_pkinase) == "try-error" && class(hmmer_kinase_tyr) == "try-error") {
    
      warning("Input HMM files are empty")
    
    kinase_positions = "Error"
    
  } else if (class(hmmer_pkinase) == "try-error" && class(hmmer_kinase_tyr) != "try-error"){
    
     colnames(hmmer_kinase_tyr) = c("target_name", "accession_species", "tlen", "domain_name", "accession_hmm", "qlen", "E-value",  "sequence_score",  "sequence_bias", "sequence_num", "seq_of", "c-Evalue",  "i-Evalue",  "domain_score",  "domain_bias",  "hmm_from",    "hmm_to",  "ali_from",    "ali_to",  "env_from",     "env_to",  "acc", "description_of_target")
     
    kinase_positions = as.data.frame(hmmer_kinase_tyr)
    kinase_positions$domain_name=str_remove(kinase_positions$domain_name,"\\..*")
    
  } else if (class(hmmer_pkinase) != "try-error" && class(hmmer_kinase_tyr) == "try-error") {
    
    colnames(hmmer_pkinase) = c("target_name", "accession_species", "tlen", "domain_name", "accession_hmm", "qlen", "E-value",  "sequence_score",  "sequence_bias", "sequence_num", "seq_of", "c-Evalue",  "i-Evalue",  "domain_score",  "domain_bias",  "hmm_from",    "hmm_to",  "ali_from",  "ali_to",  "env_from", "env_to",  "acc", "description_of_target")
    
    kinase_positions = as.data.frame(hmmer_pkinase)
    kinase_positions$domain_name=str_remove(kinase_positions$domain_name,"\\..*")
    
  } else if (class(hmmer_pkinase) != "try-error" && class(hmmer_kinase_tyr) != "try-error") {
    
  colnames(hmmer_pkinase) = c("target_name", "accession_species", "tlen", "domain_name", "accession_hmm", "qlen", "E-value",  "sequence_score",  "sequence_bias", "sequence_num", "seq_of", "c-Evalue",  "i-Evalue",  "domain_score",  "domain_bias",  "hmm_from",    "hmm_to",  "ali_from",  "ali_to",  "env_from", "env_to",  "acc", "description_of_target")
  
   
  colnames(hmmer_kinase_tyr) = c("target_name", "accession_species", "tlen", "domain_name", "accession_hmm", "qlen", "E-value",  "sequence_score",  "sequence_bias", "sequence_num", "seq_of", "c-Evalue",  "i-Evalue",  "domain_score",  "domain_bias",  "hmm_from",    "hmm_to",  "ali_from",    "ali_to",  "env_from",     "env_to",  "acc", "description_of_target")
  
  hmmer_pkinase = as.data.frame(hmmer_pkinase)
  hmmer_kinase_tyr = as.data.frame(hmmer_kinase_tyr)
  
  #hmmer_pkinase = read_domtblout(paste(hmmer_dir,input_pkinase_files[i], sep=""))

  #hmmer_kinase_tyr = read_domtblout(paste(hmmer_dir,input_pkinase_tyr_files[i],sep=""))
  
  extra_tyr = hmmer_kinase_tyr %>% filter(!(target_name %in% hmmer_pkinase$target_name))
  
  kinase_positions = rbind(hmmer_pkinase, extra_tyr)
  kinase_positions$domain_name=str_remove(kinase_positions$domain_name,"\\..*")
  
  }
  
  if (kinase_positions != "Error"){
  
  #I had a thing to make the names nicer, but eh w/e. It works weird with different inputs. 
  #gene_name = str_replace(basename(input_files[i]), "_.*","")
  gene_name = input_files[i]
  
  #kinase_start_pos = as.vector(kinase_positions[grep(gene_name,kinase_positions$domain_name),'env_from'])
  kinase_start_pos = as.vector(min(kinase_positions$env_from))
  
  #get json of selection 
  selection_json = rjson::fromJSON(file=paste(selection_dir,input_files[i],sep=""))
  
  #get number of sites being tested in each section
  gene_length = selection_json$input$`number of sites`
  ecd_length = kinase_start_pos
  kinase_length = (gene_length - ecd_length) 
  
  selection_df = as.data.frame(t(as.data.frame(selection_json$MLE$content$`0`)))
  rownames(selection_df)=c(1:nrow(selection_df))
  selection_df = cbind(selection_df, rownames(selection_df)) 
  colnames(selection_df) = c("alpha", "beta", "beta-alpha, mean posterior beta-alpha", "negative", "positive", "bayes_factor", "null1", "null2", "Position")
  #get total number of positive and negative selected sites gene wide
    
    pos_sites = selection_df %>% filter(positive > filter_cutoff)
    
    neg_sites = selection_df %>% filter(negative > filter_cutoff)
    
    #ratio
    selection_ratio = nrow(pos_sites)/nrow(neg_sites)
    
    
    ###kinase filtering section#####
    
    #because the TM domain is hard to accurately determine, the "ECD" is actually the "N-terminal of the kinase domain" 
    ecd_sites = selection_df[1:kinase_start_pos[[1]],]
    kinase_sites = selection_df[kinase_start_pos[[1]]:nrow(selection_df),]  
    
    ##ecd pos and neg
    pos_ecd = ecd_sites %>% filter(positive > filter_cutoff)
  
    neg_ecd = ecd_sites %>% filter(negative > filter_cutoff) 
    
    #kinase pos and neg
    pos_kinase = kinase_sites %>% filter(positive > filter_cutoff)
   
    neg_kinase = kinase_sites %>% filter(negative > filter_cutoff) 
    
    #get simple ratio of pos to neg selection 
    ecd_pos_neg_ratio = nrow(pos_ecd)/nrow(neg_ecd)
    kinase_pos_neg_ratio = nrow(pos_kinase)/nrow(neg_kinase)
  
  if (selection_json$MLE$content$`0`[1]!="NULL"){
    
    
    
    
    
    
   

    
    
    
    ####output section
    
    output_matrix[i,1:10] = c(gene_name, (nrow(pos_sites)), (nrow(neg_sites)), selection_ratio, (nrow(pos_ecd)), (nrow(neg_ecd)), ecd_pos_neg_ratio, (nrow(pos_kinase)), (nrow(neg_kinase)), kinase_pos_neg_ratio)
    
    #add the sites found under selection to the output matrix. Here we divide the value of each by the feature length to get a "rate of positive/negatively selected sites per site" 
    output_matrix_normalized[i,1:10] = c(gene_name, (nrow(pos_sites)/gene_length), (nrow(neg_sites)/gene_length), selection_ratio, (nrow(pos_ecd)/ecd_length), (nrow(neg_ecd)/ecd_length), ecd_pos_neg_ratio, (nrow(pos_kinase)/kinase_length), (nrow(neg_kinase)/kinase_length), kinase_pos_neg_ratio)
    
    
    
    
    
    
  } else {
    
    pos_sites = matrix(nrow=0, ncol=9)
    colnames(pos_sites) = c("alpha", "beta", "beta-alpha, mean posterior beta-alpha", "negative", "positive", "bayes_factor", "null1", "null2", "Position")
    
    neg_sites = matrix(nrow=0, ncol=9)
    colnames(neg_sites) = c("alpha", "beta", "beta-alpha, mean posterior beta-alpha", "negative", "positive", "bayes_factor", "null1", "null2", "Position")
    
    #get the average nonsyn rate and syn rate across the genes being compared
    
    syn_sites = mean(selection_df$alpha)
    
    nonsyn_sites = mean(selection_df$beta)
    
    #split the syn rates up by feature
    #ecd
    syn_ecd = mean(ecd_sites$alpha)
    nonsyn_ecd = mean(ecd_sites$beta)
    
    #kinase
    syn_kinase = mean(kinase_sites$alpha)
    nonsyn_kinase = mean(kinase_sites$beta)
    
    #put gene-wide substitution rate into matrix
    substitution_rate_matrix[i, 1:6] = c(syn_sites, nonsyn_sites, syn_ecd, nonsyn_ecd, syn_kinase, nonsyn_kinase)
    
    #put normalized gene-wide substitution rate into matrix
    substitution_rate_matrix_normalized[i, 1:6] = c(syn_sites/gene_length, nonsyn_sites/gene_length, syn_ecd/ecd_length, nonsyn_ecd/ecd_length, syn_kinase/kinase_length, nonsyn_kinase/kinase_length)
    
    
    
    output_matrix[i,1:10] = c(gene_name, 0, 0, 0, 0, 0, 0, 0, 0, 0)
     
     
    }
  
  }
  
    
  
   
  
  
  #output positively selected sites
    write.csv(pos_sites, paste(output_dir, input_files[i], "_posSelection.csv", sep=""))
  
  #output negatively selected sites
   write.csv(neg_sites, paste(output_dir, input_files[i], "_negSelection.csv", sep=""))
  
}

#output total
write.csv(output_matrix, paste(output_dir, "RLKs_ALL.csv", sep=""), row.names=FALSE)
write.csv(output_matrix_normalized, paste(output_dir, "RLKs_ALL_Normalized.csv", sep=""), row.names=FALSE)

write.csv(substitution_rate_matrix, paste(output_dir, "RLKs_ALL_Substitution_Rate.csv", sep=""), row.names=TRUE)
write.csv(substitution_rate_matrix_normalized, paste(output_dir, "RLKs_ALL_Substitution_Rate_Normalized.csv", sep=""), row.names=TRUE)

##why are the outputs different? why does one have rownames in the first column and the other has gene names as the rownames? Honestly i forgot which way I liked better so I'm doing both...
```


####for graphing selection analysis stats per OG

```{r}

library(ggplot2)
library(tidyr)
library(ggpubr)
library(dplyr)
library(rstatix)
library(multcompView)
library(dunn.test)


csv_in = read.csv("D:/Analysis_Output/hyphy/CDS/Seqkit_filtered_length/Filtered/Selection_Parsed/For_graphing.csv")

csv_in = read.csv("D:/Analysis_Output/hyphy/CDS/Dec_stuff/RLKs_only/Selection_Parsed/Sites_normalized_for_graphing.csv")

###This is for the substitution rate
csv_in = read.csv("D:/Analysis_Output/hyphy/CDS/Dec_stuff/RLKs_only/Selection_Parsed/Substitution_rate_graphing.csv")

csv_in = csv_in %>% filter(Duplication != "Singleton")


#stress_families = c("MLRR", "CRK", "SRK")

#csv_in = csv_in %>% filter(Family %in% stress_families)


csv_in = csv_in %>% filter(Family != "RLCK")

#test_df = as.data.frame(cbind(as.numeric(csv_in$Pos_ECD), csv_in$Duplication))
#colnames(test_df) = c("Pos_ECD", "Duplication")
#compare_means(Duplication ~ Pos_ECD, data = csv_in, method = "anova")

#csv_in$Neg_ECD[csv_in$Neg_ECD == 0]=1
#csv_in$PosNeg_ECD_Ratio = csv_in$Pos_ECD/csv_in$Neg_ECD

csv_in$nonsyn_syn_ratio = csv_in$NonSyn_Rate/csv_in$Syn_rate
csv_in$ecd_non_syn_ratio = csv_in$NonSyn_ECD/csv_in$Syn_ECD
csv_in$kinase_non_syn_ratio = csv_in$NonSyn_Kinase/csv_in$Syn_Kinase

gathered_csv = gather(csv_in, key="Substitution_type", value="Substitution_rate", -"OG", -"Duplication", -"Pangenome", -"Family")

###With stats
#factor if using duplication as X
csv_in$Duplication = factor(csv_in$Duplication, levels = c("WGD/Segmental", "Dispersed", "Proximal","Tandem"))
#factor for stats if using pangenoem as x alphabetical
#csv_in$Pangenome = factor(csv_in$Pangenome, levels = c("Almost Core", "Cloud","Core","Shell"))
#factor if you want a specific order. YOU MUST MAKE SURE test_stats follows this as well 
csv_in$Pangenome = factor(csv_in$Pangenome, levels = c("Core", "Almost Core", "Shell", "Cloud"))
#factor fo family
csv_in$Family = factor(csv_in$Family, levels = c("CR4L","CRK","CrRLK1L", "Extensin", "LecRLK", "LRK", "LRK10L", "LRR", "MLRR", "PR5L", "PERKL","RKF3L", "RLCK", "SRK", "URK", "WAKL"))

  ####NOTE::::: if using dunn-sidak test, i'm using a workaround hack to follow the format of rstatix.
  #####I'm just adding in the value of the p-value for plotting purposes 
test_stats = csv_in %>% pairwise_t_test(nonsyn_syn_ratio ~ Pangenome, p.adjust.method = "bonferroni")
dunn_sidak_test = dunn.test(x=csv_in$Num_Pos_Sites, g=csv_in$Duplication, method="sidak", kw=FALSE, wrap=TRUE, label=TRUE, table=TRUE, list=FALSE)
dunn_sidak_matrix = t(matrix(unlist(dunn_sidak_test[c(2:5)]), ncol=6, byrow=TRUE))
test_stats$p.adj = signif(as.numeric(dunn_sidak_matrix[,3]),3)


test_stats$xmin = c(2,2,3,1,1,1) ###add order for duplication type
test_stats$xmax = c(3,4,4,2,3,4) #add order for duplication type 

test_stats$xmin = c(2,1,1,2,3,1) ###add order for pangenome type
test_stats$xmax = c(4,2,4,3,4,3) #add order for pangenome type 

test_stats$y.position = max(csv_in$Num_Pos_Sites*1.5)

#csv_in = csv_in %>% filter(Pos_ECD >= 1)

anova_stat = aov(Num_Pos_Sites ~ Family, data = csv_in)
tukey_stat = TukeyHSD(anova_stat)
letter_df = multcompLetters4(anova_stat,tukey_stat)
dt = group_by(csv_in,Family) %>% 
  summarise(w=mean(Num_Pos_Sites), sd=sd(Num_Pos_Sites)) %>%
  arrange(desc(w))

letter_df = as.data.frame.list(letter_df$Family)
dt$Letters = letter_df$Letters

#anova_tukey_test = aov(nonsyn_syn_ratio ~ Pangenome, data = csv_in) %>% tukey_hsd(formula=nonsyn_syn_ratio ~ Pangenome, data=csv_in)
#anova_tukey_test$groups = 
#anova_tukey_test$xmin = c(1,1,1,2,2,3)
#anova_tukey_test$xmax = c(2,3,4,3,4,4)

#anova_tukey_test = anova_tukey_test %>% add_xy_position(x="Duplication")

tukey_test = csv_in %>% tukey_hsd(Num_Pos_Sites ~ Duplication) 

tukey_test = tukey_test %>% add_xy_position(x="Duplication")

#this is veryuyyyyyy specific to reorder based on the order I want them compared in
#tukey_test$xmin = c(2,2,2,4,4,1)
#tukey_test$xmax = c(4,1,3,1,3,3)

test_stats = tukey_test

anova_tukey_test$term = "Pos_ECD"
colnames(anova_tukey_test[1]) = ".y."
anova_tukey_test = 
#test_stat = group_by(csv_in, Pos_ECD) %>% summarise(w=mean(Pos_ECD), sd=sd(Pos_ECD)) %>% arrange(desc(w))
  
  



#t_test_stat = t.test(csv_in$Pos_Kinase, csv_in$Pos_ECD, paired=TRUE)

#test_stats = csv_in %>% dunn_test(Pos_ECD ~ Duplication, p.adjust.method="fdr")
#pairwise_
test_stats = test_stats %>% add_xy_position(x="Duplication")
letter_stats = multcompLetters(anova,dunn_test)

#gathered_csv = data.frame(csv_in$Gene, csv_in$Pos_ECD, csv_in$Pos_Kinase)
gathered_csv = csv_in %>% select(Gene, Pos_ECD, Pos_Kinase, Family, Duplication, Pangenome)
gathered_csv = gather(gathered_csv, key="Selection_Type", value="Sel_value", -"Gene", -"Duplication", -"Pangenome", -"Family")
gathered_csv$Pangenome = factor(gathered_csv$Pangenome, levels=c("Core", "Almost Core", "Shell", "Cloud"))
gathered_csv$Duplication = factor(gathered_csv$Duplication, levels=c("WGD/Segmental", "Dispersed", "Proximal", "Tandem"))

subset_test = gathered_csv %>% filter(test_column %in% c("Dispersed,Pos_ECD", "Dispersed,Pos_Kinase"))
dunn_sidak_test = dunn.test(x=subset_test$Sel_value, g=subset_test$test_column, method="sidak", kw=FALSE, wrap=TRUE, label=TRUE, table=TRUE, list=FALSE)
test_comparisons = c("Dispersed,Pos_ECD - Dispersed,Pos_Kinase", "WGD/Segmental,Pos_ECD - WGD/Segmental,Pos_Kinase", "Proximal,Pos_ECD - Proximal,Pos_Kinase", "Tandem,Pos_ECD - Tandem,Pos_Kinase")

test_stats = gathered_csv %>% pairwise_t_test(Sel_value ~ Selection_Type, p.adjust.method = "bonferroni")
dunn_sidak_test = dunn.test(x=gathered_csv$Sel_value, g=gathered_csv$test_column, method="sidak", kw=FALSE, wrap=TRUE, label=TRUE, table=TRUE, list=FALSE)
dunn_sidak_matrix = t(matrix(unlist(dunn_sidak_test[c(2:5)]), ncol=length(dunn_sidak_test$comparisons), byrow=TRUE))
test_stats$p.adj = signif(as.numeric(dunn_sidak_matrix[,3]),3)
test_stats$y.position = max(gathered_csv$Sel_value)


#gathered_csv = csv_in %>% select(Gene, Pos_ECD, Pos_Kinase, Duplication)
#gathered_csv = gather(gathered_csv, key="Selection_Type", value="Sel_value", -"Gene", -"Duplication")
gathered_csv$test_column = paste(gathered_csv$Duplication, gathered_csv$Selection_Type, sep=",")
gathered_csv$test_column = factor(gathered_csv$test_column, levels=c("Dispersed,Pos_ECD", "Dispersed,Pos_Kinase", "Proximal,Pos_ECD","Proximal,Pos_Kinase","Tandem,Pos_ECD","Tandem,Pos_Kinase","WGD/Segmental,Pos_ECD","WGD/Segmental,Pos_Kinase"))
gathered_csv$test_column = factor(gathered_csv$test_column, levels=c("Core,Pos_ECD", "Core,Pos_Kinase", "Almost Core,Pos_ECD","Almost Core,Pos_Kinase","Shell,Pos_ECD","Shell,Pos_Kinase","Cloud,Pos_ECD","Cloud,Pos_Kinase"))

#tandem_gathered_csv = gathered_csv %>% filter(Duplication == "Tandem")
#test_stats = gathered_csv %>% dunn.test(Substitution_rate ~ Substitution_type, method = "sidak")
test_stats = rbind(test_stats[1,],test_stats[6,], test_stats[15,], test_stats[28,])



test_stats = gathered_csv %>% pairwise_t_test(Sel_value~ test_column, p.adjust.method = "bonferroni")
#test_stats = gathered_csv %>% tukey_hsd(Sel_value ~ test_column)
test_stats = test_stats %>% add_xy_position(x="Selection_Type")   ##sometiems this doesn't work and I don't know why. Just add them manually 
added_xypos = test_stats %>% add_xy_position(x="Duplication")
test_stats = rbind(test_stats[1,],test_stats[6,], test_stats[15,], test_stats[28,])


##########Colours
duplication_palette = c("#ffaabb", "#aaaa00", "#99ddff", "#ee8866")

###############

test_stats = filter(test_stats, p.adj.signif < 0.05)


##GGPLOT NOT THE OTHER ONE
#by pangenome only
ggplot(data=csv_in, aes(y=Num_Neg_Sites, x=Duplication, fill=Pangenome)) + geom_boxplot()
#byfamily
ggplot(data=csv_in, aes(y=Num_Neg_Sites, x=Family)) + geom_boxplot()

ggplot(data=csv_in, aes(x=Pangenome, y=Pos_ECD, fill=Family)) + geom_boxplot(outlier.size=0.1) + geom_point(aes(x=Pangenome, y=Pos_ECD, fill=Family), size=0.1) + facet_wrap(~Duplication, scales="free")

ggplot(data = csv_in, aes(x=Family, y=Pos_ECD, fill=Family)) + geom_violin() + scale_fill_manual(values=c("#d563ba", "#73b638", "#9f58c0","#5fb96d","#d33d72","#48b6a5","#d84e37","#676bcf","#b5ad45","#7191cf","#de9039","#9a4c73","#5b7c39","#df859c","#a87c42", "#b85944"))

##gathered

ggboxplot(data=gathered_csv, x="Family", y="Sel_value", fill="Selection_Type", groups="Duplication") + scale_fill_manual(values=c("darkgreen", "gold")) #+ geom_point(aes(x=Family, y=Sel_value, group=Selection_Type, color=Duplication), size=0.1) + scale_color_manual(values=duplication_palette) #+ stat_pvalue_manual(test_stats, label = "p.adj", tip_length = 0, bracket.size=0.1, step.increase = 0.05)

ggboxplot(data=gathered_csv, x="test_column", y="Sel_value", fill="Selection_Type", groups="Pangenome") + stat_pvalue_manual(test_stats, label = "p.adj", tip_length = 0, bracket.size=0.1, step.increase = 0.00) + scale_fill_manual(values=c("darkgreen", "gold", "darkgreen", "gold", "darkgreen", "gold", "darkgreen", "gold"))  + scale_x_discrete(labels=c('Core','','Almost Core','','Shell','','Cloud','')) + theme(axis.text = element_text(hjust=-0.2))

ggboxplot(data=gathered_csv, x="Selection_Type", y="Sel_value", fill="Pangenome") + scale_fill_manual(values=c("#3C4142", "grey", "orange", "blue")) + stat_pvalue_manual(test_stats, label = "p.adj", tip_length = 0, bracket.size=0.1, step.increase = 0.05) 

ggboxplot(data=gathered_csv, x="Selection_Type", y="Sel_value", fill="Duplication") + scale_fill_manual(values=duplication_palette) + stat_pvalue_manual(test_stats, label = "p.adj", tip_length = 0, bracket.size=0.1, step.increase = 0.05) + scale_y_continuous(breaks=seq(0,0.15,0.05))

ggboxplot(data=gathered_csv, x="Selection_Type", y="Sel_value", fill="Family") + stat_pvalue_manual(test_stats, label = "p.adj", tip_length = 0, bracket.size=0.1, step.increase = 0.05) + scale_fill_manual(values=c("#d563ba", "#73b638", "#9f58c0","#5fb96d","#d33d72","#48b6a5","#d84e37","#676bcf","#b5ad45","#7191cf","#de9039","#9a4c73","#5b7c39","#df859c","#a87c42", "#b85944"))


###not gathered
ggboxplot(data=csv_in, x="Duplication", y="Num_Pos_Sites", fill="Pangenome") + stat_pvalue_manual(test_stats, label = "p.adj", tip_length = 0, bracket.size=0.1, step.increase = 0.1) + scale_fill_manual(values=c("#3C4142", "grey", "orange", "blue"))

ggboxplot(data=csv_in, x="Duplication", y="Num_Pos_Sites", fill="Duplication") + stat_pvalue_manual(test_stats, label = "p.adj", tip_length = 0, bracket.size=0.1, step.increase = 0.1) + scale_fill_manual(values=duplication_palette)

ggboxplot(data=csv_in, x="Family", y="Num_Pos_Sites", fill="Family", ) + stat_pvalue_manual(test_stats, label = "p.adj", tip_length = 0, bracket.size=0.1, step.increase = 0.05) + scale_fill_manual(values=c("#d563ba", "#73b638", "#9f58c0","#5fb96d","#d33d72","#48b6a5","#d84e37","#676bcf","#b5ad45","#7191cf","#de9039","#9a4c73","#5b7c39","#df859c","#a87c42", "#b85944"))

ggviolin(data=csv_in, x="Family", y="Pos_ECD", fill="Family", add="boxplot") + stat_pvalue_manual(test_stats, label = "p.adj", tip_length = 0, bracket.size=0.1, step.increase = 0.05) + scale_fill_manual(values=c("#3C4142", "grey", "orange", "blue"))

ggviolin(data=csv_in, x="Pangenome", y="ecd_non_syn_ratio", fill="Duplication", add = "boxplot", add.param=list(color="black")) + stat_pvalue_manual(test_stats, label = "p.adj", tip_length = 0, bracket.size=0.1, step.increase = 0.05) + scale_fill_manual(values=duplication_palette)

ggviolin(data=csv_in, x="Family", y="Num_Pos_Sites", fill="Family", trim=TRUE, draw.quantiles=TRUE) + scale_fill_manual(values=c("#d563ba", "#73b638", "#9f58c0","#5fb96d","#d33d72","#48b6a5","#d84e37","#676bcf","#b5ad45","#7191cf","#de9039","#9a4c73","#5b7c39","#df859c","#a87c42", "#b85944")) + geom_text(data=dt, aes(label=Letters, y=0.15) )

#ggviolin(data=csv_in, x="Family", y="Pos_ECD", fill="Family", trim=TRUE) + scale_fill_manual(values=c("#d563ba", "#73b638", "#9f58c0","#5fb96d","#d33d72","#48b6a5","#d84e37","#676bcf","#b5ad45","#7191cf","#de9039","#9a4c73","#5b7c39","#df859c","#a87c42", "#b85944"))

mean_sites = mean(csv_in$Num_Pos_Sites)
sd_sites = sd(csv_in$Num_Pos_Sites)
csv_in_test=csv_in
csv_in_test$Num_Pos_Sites[csv_in_test$Num_Pos_Sites>(mean_sites + 3*sd_sites)] = mean_sites + 3.1*sd_sites
csv_subset = select(csv_in, Family, Num_Pos_Sites)
csv_subset$Num_Pos_Sites = ((csv_subset$Num_Pos_Sites - mean_sites)/sd_sites)

mean_sites = mean(csv_subset$Num_Pos_Sites)
sd_sites = sd(csv_subset$Num_Pos_Sites)
csv_subset$Num_Pos_Sites[csv_subset$Num_Pos_Sites>(mean_sites + 3*sd_sites)] = mean_sites + 3*sd_sites

##histogram 
gghistogram(data=csv_in_test, x="Num_Pos_Sites",y="count", bins=30, merge=TRUE, alpha=1, rug=FALSE, fill="Family", position="stack", facet.by="Family", add_density=FALSE, palette=c("#d563ba", "#73b638", "#9f58c0","#5fb96d","#d33d72","#48b6a5","#d84e37","#676bcf","#b5ad45","#7191cf","#de9039","#9a4c73","#5b7c39","#df859c","#a87c42", "#b85944"))  +

  geom_vline(xintercept = mean_sites + 1*sd_sites, color="black", linetype="dashed") +
  geom_vline(xintercept = mean_sites + 2*sd_sites, color="orange", linetype="dashed") +
  geom_vline(xintercept = mean_sites + 3*sd_sites, color="blue", linetype="dashed") +
  geom_vline(xintercept = mean_sites - 1*sd_sites, color="black", linetype="dashed") +
  geom_vline(xintercept = mean_sites, color="red", linetype="dashed") +
  annotate("text", x=mean_sites + 0.8*sd_sites, y=100, color="black", label="+1*SD" ) +
  annotate("text", x=mean_sites + 1.8*sd_sites, y=100, color="orange", label="+2*SD" ) +
  annotate("text", x=mean_sites + 2.8*sd_sites, y=100, color="blue", label="+3*SD" ) +
  annotate("text", x=mean_sites - 0.8*sd_sites, y=100, color="black", label="-1*SD" ) +

  annotate("text", x=mean_sites, y=100, color="red", label=paste0("mean=",round(mean_sites, 3)) ) +
  xlab("PSS per nucleotide") +
  ylab("Bin density") +
   theme_classic()
  

ggplot(data=csv_in, aes(x=Pangenome, y=after_stat(count))) + geom_bar(aes(fill=Duplication), position='fill', width=0.69) + scale_fill_manual(values=c("#ffaabb", "#aaaa00", "#99ddff", "#ee8866" )) + theme_classic() + scale_y_continuous(labels=c("0%", "10%", "20%",'30%','40%','50%','60%','70%','80%','90%','100%'), breaks=(seq(0, 1, by=0.1)), expand = c(0, 0), limits = c(0, NA))




ggsave(paste("D:/Analysis_Output/Orthogroup_Assignment/Sep2024/Kinases/Orthofinder_out/Figures/Pos_Pangenome_ECDvsKinaseFill_PangenomeFill90conf.png", sep=""),height=15, width=15, dpi=600)

ggsave(paste("D:/Analysis_Output/Orthogroup_Assignment/Sep2024/Kinases/Orthofinder_out/Figures/Kinase_Syn_nonsyn_rate_Duplication_normalized.png", sep=""),height=10, width=15, dpi=600)

ggsave(paste("D:/Analysis_Output/Orthogroup_Assignment/Sep2024/Kinases/Orthofinder_out/Results_Dec9/Figures/Alternate_colours_figures/ECD_vs_KINASE_PangenomeSplit.png", sep=""),height=10, width=15, dpi=600)

ggsave(paste("D:/Analysis_Output/Orthogroup_Assignment/Sep2024/Kinases/Orthofinder_out/Results_Dec9/Figures/NumPosSites_Histogram_z-scores_include_1_sd_FACETED.png", sep=""),height=10, width=20, dpi=600)

```




##amino acids under selection

```{r}

#library(rjson)
library(jsonlite)
library(dplyr)
library(stringr)
library(phylotools)

selection_out_folder = "D:/Analysis_Output/hyphy/CDS/Dec_stuff/RLKs_only/Selection_Parsed/"
alignment_folder = "D:/Analysis_Output/hyphy/CDS/Dec_stuff/RLKs_only/Translated/"
output_folder = "D:/Analysis_Output/hyphy/CDS/Dec_stuff/RLKs_only/Selection_Parsed/Common_AAs/"

pos_sel_files = list.files(selection_out_folder, pattern = "*posSelection.csv")
neg_sel_files = list.files(selection_out_folder, pattern = "*negSelection.csv")
MSA_files = list.files(alignment_folder, pattern = "*.fasta")




for (i in 1:length(MSA_files)) {
  
  
  #alignment_file = read.fasta(paste(alignment_folder, MSA_files[i], sep=""))
#alignment_file = read_fasta(paste(alignment_folder, MSA_files[i], sep=""), type="AA")
alignment_file = read.fasta(paste(alignment_folder, MSA_files[i], sep=""))

pos_selection_in = read.csv(paste(selection_out_folder, pos_sel_files[i], sep=""))

print(paste("working on ", selection_out_folder, pos_sel_files[i], sep=""))

AA_positions = pos_selection_in$Position

  sites_list = vector(mode="list", length=length(AA_positions))

names(sites_list) = AA_positions

if (length(AA_positions) >0){

  for (j in 1:length(AA_positions)) {
  
  #AAs_position_vector = seq_extract_position(alignment_file, AA_positions[j], AA_positions[j])
  AAs_position_vector = substr(alignment_file$seq.text, AA_positions[j], AA_positions[j])
  
   
  
  
  
  #unique_AAs = unique(AAs_position_vector)
  
  #AAs_under_selection = matrix(ncol=2, nrow=length(unique_AAs))
  #AAs_under_selection[,1] = unique_AAs
  
  sites_list[[j]]= table(AAs_position_vector)
  
   
  }
  
#lapply(sites_list, write, paste(output_folder,pos_sel_files[i], "_AAs.txt", sep=""),append=TRUE,ncolumns=1000)

#json_out = jsonlite::toJSON(sites_list, pretty=TRUE, auto_unbox=TRUE, force=TRUE)
#json_out = rjson::toJSON(sites_list, indent = TRUE)


sink(paste(output_folder,pos_sel_files[i], "_AAs.txt", sep=""))
print(sites_list)
sink()

} else {
  warning("No sites under selection")
  sink(paste(output_folder,pos_sel_files[i], "_AAs.txt", sep=""))
print(sites_list)
sink()
} 

}
#AA_
  
  #thingy = "ATGTCGG"

#thingyAA = seq_translate(dna(thingy))


```

####plot representative genes from each OG and put on them domains and location of POS or NEG selected sites
```{r}
#library(jsonlite)
library(dplyr)
#library(stringr)
library(phylotools)
library(ggplot2)
library(rhmmer)

consensus_fasta_in = read.fasta("D:/Analysis_Output/HMMER/RLK_OG_Consensus/RLK_OGs_Consensus.fasta")
gene_lengths = consensus_fasta_in 
gene_lengths$seq.text = nchar(consensus_fasta_in$seq.text)

selection_out_folder="D:/Analysis_Output/hyphy/CDS/Selection_Parsed/"
alignment_folder="D:/Analysis_Output/hyphy/CDS/Translate_CDS/"
output_folder = "D:/Analysis_Output/hyphy/CDS/Selection_Parsed/Common_AAs/"

hmmscan_input_folder = 

pos_sel_files = list.files(selection_out_folder, pattern = "*posSelection.csv")
neg_sel_files = list.files(selection_out_folder, pattern = "*negSelection.csv")

```






